source/opt/unify_const_pass.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "unify_const_pass.h"
+
+#include <unordered_map>
+#include <utility>
+
+#include "def_use_manager.h"
+#include "make_unique.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+// The trie that stores a bunch of result ids and performs equivalent result id
+// looking up with a given instruction.
+class ResultIdTrie {
+ public:
+  ResultIdTrie() : root_(new Node) {}
+
+  // For a given instruction, extracts its opcode, type id and operand words
+  // as an array of keys, looks up the trie to find a result id which is stored
+  // with the same opcode, type id and operand words. If none of such result id
+  // is found, creates a trie node with those keys, stores the instruction's
+  // result id and returns that result id. If an existing result id is found,
+  // returns the existing result id.
+  uint32_t LookupEquivalentResultFor(const ir::Instruction& inst) {
+    auto keys = GetLookUpKeys(inst);
+    auto* node = root_.get();
+    for (uint32_t key : keys) {
+      node = node->GetOrCreateTrieNodeFor(key);
+    }
+    if (node->result_id() == 0) {
+      node->SetResultId(inst.result_id());
+    }
+    return node->result_id();
+  }
+
+ private:
+  // The trie node to store result ids.
+  class Node {
+   public:
+    using TrieNodeMap =
+        std::unordered_map<uint32_t, std::unique_ptr<Node>>;
```

Comment:
I suspect a more specialized data structure would perform better.  But we have no tuning set, so let's not worry about it now. :-)